File: game/plane.py
from dataclasses import dataclass

from game.plane_data import PlaneStats, PlaneType, Vector

@dataclass
class Plane:
    id: str
    team: str
    type: PlaneType
    position: Vector
    angle: float
    health: int
    stats: PlaneStats

    def deserialize(blob: object) -> "Plane":
        try:
            plane = Plane(
                blob["id"],
                blob["team"],
                PlaneType[blob["type"]],
                Vector.deserialize(blob["position"]),
                blob["angle"],
                blob["health"],
                PlaneStats.deserialize(blob["stats"])
            )
        except:
            print("Failed to validate plane json")
            raise

        return plane

File: engine/content/src/plane/plane.ts
import { PLANE_STATS } from "../config"
import { PlaneStats, PlaneType } from "./data"
import { Vector } from "./vector"

export type PlaneId = string

export class Plane {
    public health: number
    public stats: PlaneStats
    constructor(
        readonly id: PlaneId,
        readonly team: number,
        readonly type: PlaneType,
        public position: Vector,
        public angle: number // Angle that faces [0, 360) so 0 = East, 90 = North, etc.
    ) {
        this.stats = PLANE_STATS[type]
        this.health = this.stats.maxHealth
    }
}
